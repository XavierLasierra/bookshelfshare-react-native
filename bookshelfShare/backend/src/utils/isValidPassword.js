const bcrypt = require("bcrypt");

function isValidPassword(password) {
  return bcrypt.compareSync(password, this.password);
}

module.exports = { isValidPassword };
