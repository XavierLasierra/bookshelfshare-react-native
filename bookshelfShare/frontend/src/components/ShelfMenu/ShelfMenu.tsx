import React, { useEffect, useState } from 'react';
import {
  Menu, MenuOptions, MenuTrigger, MenuOption, renderers
} from 'react-native-popup-menu';
import { View, Text } from 'react-native';
import { useDispatch, useSelector } from 'react-redux';

import { addToShelf } from '../../redux/actions/userShelves.creator';

import styles from './shelfMenu.styles';
import globalStyles from '../../styles/global.styles';

export default function ShelfMenu({
  navigation, logo, bookIsbn, token, refreshToken
}: any) {
  const dispatch = useDispatch();
  const shelves = useSelector((store: any) => store.userShelves);
  const [markShelfName, setMarkShelfName] = useState('Add to shelf');
  const [deleteFromShelf, setDeleteFromShelf] = useState('');
  const [shelfLocation, setShelfLocation] = useState(null);

  function findInUserShelves() {
    let name = 'Add to shelf';
    setDeleteFromShelf('');
    setShelfLocation(null);
    shelves.forEach((shelf: any) => {
      const foundBook = shelf.books
        .find(({ bookIsbn: shelfBookIsbn }: any) => shelfBookIsbn === bookIsbn);
      if (foundBook) {
        // eslint-disable-next-line no-underscore-dangle
        setDeleteFromShelf(shelf._id);
        setShelfLocation(foundBook.customInformation?.location);
        name = shelf.name;
      }
    });
    return name;
  }

  useEffect(() => {
    setMarkShelfName(findInUserShelves());
  }, [shelves]);

  function handleShelfSelect(value: string) {
    return value === 'delete'
      ? dispatch(addToShelf(
        deleteFromShelf,
        // eslint-disable-next-line no-underscore-dangle
        '',
        bookIsbn,
        '',
        token,
        refreshToken
      ))
      : navigation.push('AddToShelf',
        {
          shelf: shelves.find(({ _id }: any) => _id === value),
          deleteFromShelf,
          bookIsbn,
          logo
        });
  }

  return (
    <View>
      <Menu
        onSelect={handleShelfSelect}
        renderer={renderers.NotAnimatedContextMenu}
      >
        <MenuTrigger>
          <Text style={globalStyles.markButton}>{markShelfName}</Text>
        </MenuTrigger>
        <MenuOptions>
          {shelves.map(({ _id, name }: any) => (
            <MenuOption
              key={_id}
              value={_id}
              text={name}
              style={globalStyles.menuOption}
            />
          ))}
          <MenuOption
            value="delete"
            text="Delete from shelf"
            style={globalStyles.menuOption}
          />
        </MenuOptions>
      </Menu>
      {shelfLocation && (
      <View style={styles.shelfLocationContainer}>
        <Text style={styles.shelfLocationText}>
          Row:
          {shelfLocation[0] + 1}
        </Text>
        <Text style={styles.shelfLocationText}>
          Col:
          {shelfLocation[1] + 1}
        </Text>
      </View>
      )}
    </View>
  );
}
