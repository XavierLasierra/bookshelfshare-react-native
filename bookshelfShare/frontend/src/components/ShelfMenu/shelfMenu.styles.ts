import {StyleSheet} from 'react-native';
import stylesConstants from '../../styles/styles.constants';

export default StyleSheet.create({
  shelfLocationContainer: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    paddingHorizontal: 20,
  },
  shelfLocationText: {
    fontSize: stylesConstants.fontSize.mid,
    fontWeight: '500',
  },
});
