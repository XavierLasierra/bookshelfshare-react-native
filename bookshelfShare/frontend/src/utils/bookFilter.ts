export function bookFilter(text: string, books: any) {
  return books.reduce((acc: any, book: any) => {
    if (book.title.toLowerCase().includes(text.toLowerCase())) return [...acc, book];
    if (book.publisher.toLowerCase().includes(text.toLowerCase())) return [...acc, book];
    if (book.isbn.ISBN13.toLowerCase().includes(text.toLowerCase())) return [...acc, book];
    if (book.isbn.ISBN10.toLowerCase().includes(text.toLowerCase())) return [...acc, book];
    if (book.authors.some((author: string) => author.toLowerCase()
      .includes(text.toLowerCase()))) return [...acc, book];

    return acc;
  }, []);
}

export function bookShelfListFilter(text: string, books: any) {
  return books.reduce((acc: any, book: any) => {
    if (book.bookData.title.toLowerCase().includes(text.toLowerCase())) return [...acc, book];
    if (book.bookData.publisher.toLowerCase().includes(text.toLowerCase())) return [...acc, book];
    if (book.bookData.isbn.ISBN13.toLowerCase().includes(text.toLowerCase())) return [...acc, book];
    if (book.bookData.isbn.ISBN10.toLowerCase().includes(text.toLowerCase())) return [...acc, book];
    if (book.bookData.authors.some((author: string) => author.toLowerCase()
      .includes(text.toLowerCase()))) return [...acc, book];

    return acc;
  }, []);
}

export function bookShelfLocationFilter(filterLocation: number[], books: any) {
  return books.filter(({ customInformation: { location } }: any) => JSON.stringify(location)
  === JSON.stringify(filterLocation));
}
