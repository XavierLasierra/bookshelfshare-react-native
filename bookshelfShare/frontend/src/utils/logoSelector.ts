import BookIcon from '../assets/bookIcon.svg';
import SearchIcon from '../assets/searchIcon.svg';
import ProfileIcon from '../assets/profileIcon.svg';

export default function logoSelector(logoName: string) {
  switch (logoName) {
    case 'SearchIcon':
      return SearchIcon;
    case 'BookIcon':
      return BookIcon;
    case 'ProfileIcon':
      return ProfileIcon;
    default:
      return BookIcon;
  }
}
