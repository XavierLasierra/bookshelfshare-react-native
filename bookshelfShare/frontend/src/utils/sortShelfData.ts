export default function sortShelfData(shelf: any) {
  return shelf?.books.reduce((acc: any, book: any) => {
    const foundShelfBox = acc.find(({ location }: any) => JSON.stringify(location)
         === JSON.stringify(book?.customInformation?.location));
    if (foundShelfBox) {
      foundShelfBox.books.push(book.bookData);
    } else {
      acc.push({
        location: book?.customInformation?.location,
        books: [book.bookData]
      });
    }
    return acc;
  }, []);
}
