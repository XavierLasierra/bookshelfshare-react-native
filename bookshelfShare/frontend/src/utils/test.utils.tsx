import React from 'react';
import { render as rtlRender } from '@testing-library/react-native';
import { Provider } from 'react-redux';

import configureStore from '../redux/store';

function render(
  component: any,
  initialState: any
) {
  const store = configureStore(initialState);

  function Wrapper({ children }: any) {
    return (
      <Provider store={store}>
        {children}
      </Provider>
    );
  }
  return rtlRender(component, { wrapper: Wrapper });
}

// re-export everything
export * from '@testing-library/react-native';
// override render method
export { render };
