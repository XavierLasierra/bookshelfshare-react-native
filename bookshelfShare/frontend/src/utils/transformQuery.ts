interface Query {
    isbn?: string,
    inauthor?: string,
    intitle?: string,
    inpublisher?: string
}

export default function transformQuery(query: Query):string {
  const queryEntries = Object.entries(query);
  const url = queryEntries
    .reduce((acc, queryElement, index) => `${acc}${queryElement[0]}=${queryElement[1]}${index < queryEntries.length - 1 ? '+' : ''}`, '');
  return `${url}&key=${process.env.GOOGLE_API_KEY}`;
}
