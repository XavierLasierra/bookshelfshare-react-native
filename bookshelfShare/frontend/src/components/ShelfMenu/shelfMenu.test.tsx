import React from 'react';
import {MenuProvider} from 'react-native-popup-menu';
import ShelfMenu from './ShelfMenu';
import {render} from '../../utils/test.utils';
import shelfMock from '../../mocks/shelf.mock';

const navigation = {
  push: jest.fn(),
};

describe('Given a ShelfMenu component', () => {
  describe('When it is rendered', () => {
    describe('And shelves does not include bookIsbn', () => {
      test('Then should match the snapshot', () => {
        const initialState = {userShelves: [shelfMock]};
        const screen = render(
          <MenuProvider>
            <ShelfMenu navigation={navigation} bookIsbn="0" />
          </MenuProvider>,
          initialState,
        );
        expect(screen).toMatchSnapshot();
      });
    });
    describe('And shelves does not include bookIsbn', () => {
      test('Then should match the snapshot', () => {
        const initialState = {userShelves: [shelfMock]};
        const screen = render(
          <MenuProvider>
            <ShelfMenu navigation={navigation} bookIsbn="1" />
          </MenuProvider>,
          initialState,
        );
        expect(screen).toMatchSnapshot();
      });
    });
  });
});
