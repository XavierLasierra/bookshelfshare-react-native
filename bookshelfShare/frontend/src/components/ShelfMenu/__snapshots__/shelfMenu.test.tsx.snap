// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given a ShelfMenu component When it is rendered And shelves does not include bookIsbn Then should match the snapshot 1`] = `
<View
  onLayout={[Function]}
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Array [
        Object {
          "flex": 1,
        },
        undefined,
        undefined,
        undefined,
      ]
    }
  >
    <View>
      <View>
        <View
          collapsable={false}
        >
          <View
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
          >
            <View
              ctx={
                Object {
                  "menuActions": Object {
                    "_getOpenedMenu": [Function],
                    "_notify": [Function],
                    "closeMenu": [Function],
                    "isMenuOpen": [Function],
                    "openMenu": [Function],
                    "toggleMenu": [Function],
                  },
                  "menuRegistry": Object {
                    "getAll": [Function],
                    "getMenu": [Function],
                    "setOptionsCustomStyles": [Function],
                    "subscribe": [Function],
                    "unsubscribe": [Function],
                    "updateLayoutInfo": [Function],
                  },
                }
              }
              style={
                Array [
                  undefined,
                  undefined,
                ]
              }
            >
              <Text
                style={
                  Object {
                    "backgroundColor": "#FFFFFF",
                    "borderColor": "#ECECEC",
                    "borderRadius": 10,
                    "borderWidth": 1,
                    "height": 40,
                    "lineHeight": 40,
                    "textAlign": "center",
                    "width": 150,
                  }
                }
              >
                Add to shelf
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
  <RCTSafeAreaView
    emulateUnlessSupported={true}
    pointerEvents="box-none"
    style={
      Object {
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  >
    <View
      collapsable={false}
      onLayout={[Function]}
      pointerEvents="box-none"
      style={
        Object {
          "flex": 1,
        }
      }
    />
  </RCTSafeAreaView>
</View>
`;

exports[`Given a ShelfMenu component When it is rendered And shelves does not include bookIsbn Then should match the snapshot 2`] = `
<View
  onLayout={[Function]}
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Array [
        Object {
          "flex": 1,
        },
        undefined,
        undefined,
        undefined,
      ]
    }
  >
    <View>
      <View>
        <View
          collapsable={false}
        >
          <View
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
          >
            <View
              ctx={
                Object {
                  "menuActions": Object {
                    "_getOpenedMenu": [Function],
                    "_notify": [Function],
                    "closeMenu": [Function],
                    "isMenuOpen": [Function],
                    "openMenu": [Function],
                    "toggleMenu": [Function],
                  },
                  "menuRegistry": Object {
                    "getAll": [Function],
                    "getMenu": [Function],
                    "setOptionsCustomStyles": [Function],
                    "subscribe": [Function],
                    "unsubscribe": [Function],
                    "updateLayoutInfo": [Function],
                  },
                }
              }
              style={
                Array [
                  undefined,
                  undefined,
                ]
              }
            >
              <Text
                style={
                  Object {
                    "backgroundColor": "#FFFFFF",
                    "borderColor": "#ECECEC",
                    "borderRadius": 10,
                    "borderWidth": 1,
                    "height": 40,
                    "lineHeight": 40,
                    "textAlign": "center",
                    "width": 150,
                  }
                }
              >
                Shelf
              </Text>
            </View>
          </View>
        </View>
      </View>
      <View
        style={
          Object {
            "flexDirection": "row",
            "justifyContent": "space-between",
            "paddingHorizontal": 20,
          }
        }
      >
        <Text
          style={
            Object {
              "fontSize": 16,
              "fontWeight": "500",
            }
          }
        >
          Row:
          1
        </Text>
        <Text
          style={
            Object {
              "fontSize": 16,
              "fontWeight": "500",
            }
          }
        >
          Col:
          1
        </Text>
      </View>
    </View>
  </View>
  <RCTSafeAreaView
    emulateUnlessSupported={true}
    pointerEvents="box-none"
    style={
      Object {
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  >
    <View
      collapsable={false}
      onLayout={[Function]}
      pointerEvents="box-none"
      style={
        Object {
          "flex": 1,
        }
      }
    />
  </RCTSafeAreaView>
</View>
`;
